tcmur_cmd_handler.c

```diff
@@ -144,6 +144,124 @@ static int write_work_fn(struct tcmu_device *dev,
 				block_size * tcmu_get_lba(cmd->cdb));
 }
 
+static void handle_write_same_16_null_cbk(struct tcmu_device *dev,
+					  struct tcmulib_cmd *cmd,
+					  int ret)
+{
+}
+
+static int write_same_16_work_fn(struct tcmu_device *dev,
+				 struct tcmulib_cmd *cmd)
+{
+	struct tcmur_handler *rhandler = tcmu_get_runner_handler(dev);
+	uint32_t block_size = tcmu_get_dev_block_size(dev);
+	uint32_t lba_cnt = tcmu_get_xfer_length(cmd->cdb);
+	uint64_t start_lba = tcmu_get_lba(cmd->cdb);
+	uint64_t end_lba;
+	uint32_t max_lbas, copy_lbas;
+	uint32_t max_sectors;
+	int ret;
+
+	end_lba = start_lba + lba_cnt;
+
+	max_sectors = tcmu_get_attribute(dev, "hw_max_sectors");
+	max_lbas = min(max_sectors, ((uint32_t)(~0U)));
+
+	tcmu_dbg("Start lba: %llu Number of lba:: %hu, end lba: %llu\n",
+		 start_lba, lba_cnt, end_lba);
+
+	/*
+	 * Write contents of the logical block data(from the Data-Out Buffer)
+	 * to each LBA in the specified LBA range.
+	 */
+	while(start_lba < end_lba) {
+		copy_lbas = min(max_lbas, lba_cnt);
+
+		if (start_lba + copy_lbas >= end_lba)
+			cmd->done = handle_generic_cbk;
+
+		ret = rhandler->write(dev, cmd, cmd->iovec, cmd->iov_cnt,
+				      tcmu_iovec_length(cmd->iovec, cmd->iov_cnt),
+				      block_size * copy_lbas);
+		if (ret)
+			goto out;
+
+		start_lba += copy_lbas;
+		lba_cnt -= copy_lbas;
+	}
+
+out:
+	return ret;
+}
+
+static int handle_write_same_16(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+{
+	uint8_t *cdb = cmd->cdb;
+	uint8_t *sense = cmd->sense_buf;
+	uint32_t blocks = tcmu_get_xfer_length(cdb);
+	uint32_t block_size = tcmu_get_dev_block_size(dev);
+	uint64_t start_lba = tcmu_get_lba(cmd->cdb);
+	uint64_t dev_last_lba = tcmu_get_dev_num_lbas(dev);
+
+	if (cmd->iov_cnt != 1 || cmd->iovec->iov_len != block_size) {
+		tcmu_err("Illegal Data-Out: iov_cnt %u length: %u\n",
+			 cmd->iov_cnt, cmd->iovec->iov_len);
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_CDB,
+					   NULL);
+	}
+
+	/*
+	 * From sbc4r13, section 5.50 WRITE SAME (16) command
+	 *
+	 * A write same (WSNZ) bit has beed set to one, so the device server
+	 * won't support a value of zero here.
+	 */
+	if (!blocks) {
+		tcmu_err("The WSNZ = 1 & WRITE_SAME blocks = 0 is not supported!\n");
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_CDB,
+					   NULL);
+	}
+
+	/*
+	 * The MAXIMUM WRITE SAME LENGTH field in Block Limits VPD page (B0h)
+	 * limit the maximum block number for the WRITE SAME.
```

Comment:
But the check_lba_and_length() couldn't be used here.
For the Write Same commands and for now the iov_cnt will be 1, and data_length in Data Out buffer will be only on block size (?). The VPD_MAX_WRITE_SAME_LENGTH is num_lbas/blocks it intend to write to.